File: web/src/vm.ts
type LynxMachine = {
  instructions: Uint8Array;
  data: Uint8Array;
  initialDataCopy: Uint8Array;
  pc: number; // u8
  // In TS we can access these be doing an array access like so: lynxMachine["r0"]. Interesting.
  r0: number;
  r1: number;
  r2: number;
  r3: number;
  cyclesExecuted: number;
};

const lynxMachine: LynxMachine = {
  instructions: new Uint8Array(256),
  data: new Uint8Array(256),
  initialDataCopy: new Uint8Array(256),
  pc: 0,
  r0: 0,
  r1: 0,
  r2: 0,
  r3: 0,
  cyclesExecuted: 0,
};

// VM execution state
let vmRunning = false;
let vmTimeoutId: number | null = null;

function startVM(): void {
  if (vmRunning) return;

  vmRunning = true;
  updateVMButtons();

  const executeInstruction = () => {
    if (!vmRunning) return;

    const halted = executeNextInstruction(lynxMachine);
    render();

    if (halted) {
      stopVM();
      console.log("VM halted");
      return;
    }

    if (vmRunning) {
      //executeInstruction();
      vmTimeoutId = setTimeout(executeInstruction, 1000 / vmSpeed);
    }
  };

  executeInstruction();
}

function stopVM(): void {
  if (!vmRunning) return;

  vmRunning = false;

  if (vmTimeoutId !== null) {
    clearTimeout(vmTimeoutId);
    vmTimeoutId = null;
  }

  updateVMButtons();
  render();
}

function updateVMButtons(): void {
  const startBtn = document.getElementById("start-btn") as HTMLButtonElement;
  const stepBtn = document.getElementById("step-btn") as HTMLButtonElement;

  if (startBtn) {
    if (vmRunning) {
      startBtn.textContent = "Stop";
      startBtn.style.backgroundColor = "#333";
      startBtn.style.color = "#fff";
    } else {
      startBtn.textContent = "Start";
      startBtn.style.backgroundColor = "";
      startBtn.style.color = "";
    }
    startBtn.disabled = false;
  }
  if (stepBtn) stepBtn.disabled = vmRunning;
}

// TODO: Move to isa.ts?
function dissasembleInstruction(inst: number): string {
  const opcode = (inst & 0b11110000) >> 4;
  const imm = inst & 0b00001111;
  let signedImm = imm;
  const reg1Idx = (inst & 0b00001100) >> 2;
  const reg2Idx = inst & 0b00000011;
  const reg1 = regNames[reg1Idx];
  const reg2 = regNames[reg2Idx];

  const instruction: Instruction | undefined = (
    Object.values(InstructionInfo) as Instruction[]
  ).find((info) => info.encoding === opcode && !info.is_directive);

  if (!instruction) return "?";

  switch (instruction.operand_kind) {
    case OperandKind.Reg:
      return `${instruction.name} ${reg1}`;
    case OperandKind.RegReg:
      return `${instruction.name} ${reg1} ${reg2}`;
    case OperandKind.Imm:
      return `${instruction.name} ${imm}`;
    default:
      return instruction.name;
  }
}

function executeNextInstruction(lynxMachine: LynxMachine): boolean {
  const inst = lynxMachine.instructions[lynxMachine.pc];

  // Decode
  const opcode = (inst & 0b11110000) >> 4;
  // Least significant 4 bits either encode an immediate or one or two registers
  const imm = inst & 0b00001111;
  let signedImm = imm;
  if (imm & 0x8) signedImm = imm - 0x10;
  // As register
  const reg1Idx = (inst & 0b00001100) >> 2;
  const reg2Idx = inst & 0b00000011;
  const reg1 = regNames[reg1Idx];
  const reg2 = regNames[reg2Idx];

  // Jump instructions sets this to false
  let advancePC = true;

  switch (opcode) {
    // Arithmetic
    case InstructionInfo[InstructionKind.Add].encoding:
      lynxMachine.r0 = (lynxMachine[reg1] + lynxMachine[reg2]) & 0xff;
      break;
    case InstructionInfo[InstructionKind.Sub].encoding:
      lynxMachine.r0 = (lynxMachine[reg1] - lynxMachine[reg2]) & 0xff;
      break;
    case InstructionInfo[InstructionKind.Inc].encoding:
      lynxMachine[reg1] = (lynxMachine[reg1] + 1) & 0xff;
      break;
    case InstructionInfo[InstructionKind.Dec].encoding:
      lynxMachine[reg1] = (lynxMachine[reg1] - 1 + 256) & 0xff;
      break;
    case InstructionInfo[InstructionKind.Ge].encoding:
      lynxMachine.r0 = Number(lynxMachine[reg1] > lynxMachine[reg2]);
      break;
    case InstructionInfo[InstructionKind.Le].encoding:
      lynxMachine.r0 = Number(lynxMachine[reg1] < lynxMachine[reg2]);
      break;
    case InstructionInfo[InstructionKind.Shift].encoding:
      const signBit = (imm & 0b1000) >> 3;
      const shiftAmount = imm & 0b0111;
      if (signBit) {
        lynxMachine.r0 = (lynxMachine.r0 >> -signedImm) & 0xff;
      } else {
        lynxMachine.r0 = (lynxMachine.r0 << shiftAmount) & 0xff;
      }
      break;
    case InstructionInfo[InstructionKind.Ali].encoding:
      lynxMachine.r0 =
        (lynxMachine.r0 + ((lynxMachine.r0 + signedImm) & 0b1111)) & 0xff;
      break;
    case InstructionInfo[InstructionKind.Li].encoding:
      lynxMachine.r0 = imm;
      break;
    // Data movement
    case InstructionInfo[InstructionKind.Mv].encoding:
      lynxMachine[reg1] = lynxMachine[reg2];
      break;
    case InstructionInfo[InstructionKind.Load].encoding:
      lynxMachine[reg1] = lynxMachine.data[lynxMachine[reg2]];
      break;
    case InstructionInfo[InstructionKind.Store].encoding:
      lynxMachine.data[lynxMachine[reg2]] = lynxMachine[reg1];
      break;
    // Jumps
    case InstructionInfo[InstructionKind.Jmp].encoding:
      // Interpret imm as 4-bit signed (-8 to 7)
      lynxMachine.pc = (lynxMachine.pc + signedImm) & 0xff;
      advancePC = false;
      break;
    case InstructionInfo[InstructionKind.Jiz].encoding:
      if (lynxMachine[reg2] === 0) {
        lynxMachine.pc = (lynxMachine.pc + lynxMachine[reg1]) & 0xff;
        advancePC = false;
      }
      break;
    case InstructionInfo[InstructionKind.Jaiz].encoding:
      if (lynxMachine[reg2] === 0) {
        lynxMachine.pc = lynxMachine[reg1];
        advancePC = false;
      }
      break;

    case InstructionInfo[InstructionKind.Halt].encoding:
      // Perhaps we should set a flag or signal something
      return true;
    default:
      alert("Could not decode instruction :-(");
      break;
  }

  if (advancePC) {
    lynxMachine.pc = (lynxMachine.pc + 1) & 0xff;
  }

  lynxMachine.cyclesExecuted++;

  return false;
}

File: chisel/src/main/scala/lynx/sw/VM.scala
package lynx.sw

case class LynxMachine(
  instructions: Array[Byte],
  data: Array[Byte],
  var pc: Byte,
  var registers: Array[Byte],  // There are 4 registers
  var cyclesExecuted: Int
) {
  // Returns true on a halt
  def step(): Boolean = {
    val instr = instructions(pc & 0xFF)
    val opcode = (instr >> 4) & 0xF
    val operand = instr & 0xF
    val instruction = Instruction.byOpcode(opcode)
    val pcModified = executeInstruction(instruction, operand)
    if (!pcModified) {
      pc = ((pc + 1) & 0xFF).toByte
    }
    cyclesExecuted += 1
    return instruction == Instruction.Halt
  }
  
  private def executeInstruction(instruction: Instruction, operand: Int): Boolean = {
    instruction.kind match {
      case OperandKind.NoneKind =>
        executeNoneInstruction(instruction)
      case OperandKind.Reg =>
        val reg = operand & 0x3
        executeRegInstruction(instruction, reg)
      case OperandKind.RegReg =>
        val op1 = (operand >> 2) & 0x3
        val op2 = operand & 0x3
        executeRegRegInstruction(instruction, op1, op2)
      case OperandKind.Imm =>
        executeImmInstruction(instruction, operand)
    }
  }
  
  private def executeNoneInstruction(instruction: Instruction): Boolean = {
    instruction match {
      case Instruction.Halt =>
        false
    }
  }
  
  private def executeRegInstruction(instruction: Instruction, reg: Int): Boolean = {
    instruction match {
      case Instruction.Inc =>
        registers(reg) = ((registers(reg) + 1) & 0xFF).toByte
        false
      case Instruction.Dec =>
        registers(reg) = ((registers(reg) - 1) & 0xFF).toByte
        false
    }
  }
  
  private def executeRegRegInstruction(instruction: Instruction, op1: Int, op2: Int): Boolean = {
    instruction match {
      case Instruction.Add =>
        registers(0) = ((registers(op1) + registers(op2)) & 0xFF).toByte
        false
      case Instruction.Sub =>
        registers(0) = ((registers(op1) - registers(op2)) & 0xFF).toByte
        false
      case Instruction.Ge =>
        registers(0) = if (registers(op1) >= registers(op2)) 1 else 0
        false
      case Instruction.Le =>
        registers(0) = if (registers(op1) <= registers(op2)) 1 else 0
        false
      case Instruction.Mv =>
        registers(op1) = registers(op2)
        false
      case Instruction.Load =>
        val address = registers(op2)
        registers(op1) = data(address)
        false
      case Instruction.Store =>
        val address = registers(op2)
        data(address) = registers(op1)
        false
      case Instruction.Jiz =>
        if (registers(op2) == 0) {
          pc = ((pc + registers(op1)) & 0xFF).toByte
          true
        } else {
          false
        }
      case Instruction.Jaiz =>
        if (registers(op2) == 0) {
          pc = registers(op1)
          true
        } else {
          false
        }
    }
  }
  
  private def executeImmInstruction(instruction: Instruction, imm: Int): Boolean = {
    instruction match {
      case Instruction.Shift =>
        val shiftAmount = if (imm > 7) -(16 - imm) else imm
        if (shiftAmount >= 0) {
          registers(0) = ((registers(0) << shiftAmount) & 0xFF).toByte
        } else {
          registers(0) = (registers(0) >> (-shiftAmount)).toByte
        }
        false
      case Instruction.Ali =>
        registers(0) = (registers(0) & imm).toByte
        false
      case Instruction.Li =>
        registers(0) = imm.toByte
        false
      case Instruction.Jmp =>
        pc = imm.toByte
        true
    }
  }
}